Chemical/CAS mapping - admin/chemical/* endpoints

mcumcu opened the pull request:
Add endpoints to support view, edit, and add CAS number mappings in the admin app. Add specs to cover Api::Admin::ChemicalsController

### Chemical routes:
```
GET    /admin/chemicals/:chemical_id
GET    /admin/chemicals/:chemical_id/map
POST   /admin/chemicals/:chemical_id/map/:goods_nomenclature_item_id
PATCH  /admin/chemicals/:chemical_id/map/:goods_nomenclature_sid(/:goods_nomenclature_item_id)
PUT    /admin/chemicals/:chemical_id/map/:goods_nomenclature_sid(/:goods_nomenclature_item_id)
DELETE /admin/chemicals/:chemical_id/map/:goods_nomenclature_sid
GET    /admin/chemicals/:chemical_id/chemical_names
GET    /admin/chemicals/:chemical_id/chemical_names/:chemical_name_id
PATCH  /admin/chemicals/:chemical_id/chemical_names
PUT    /admin/chemicals/:chemical_id/chemical_names
PATCH  /admin/chemicals/:chemical_id(/:chemical_name_id/:new_chemical_name)
PUT    /admin/chemicals/:chemical_id(/:chemical_name_id/:new_chemical_name)
GET    /admin/chemicals
POST   /admin/chemicals
```

Trelllo: https://trello.com/c/8WjkHQlC/868-admin-ui-to-view-edit-and-add-cas-number-mappings
Asana: https://app.asana.com/0/1199180449660757/1198723049518172/f

Commit:
✨ Chemical/CAS mapping - admin/chemical/* endpoints

Add endpoints to view, edit, and add CAS number mappings in the admin app. Add specs to cover Api::Admin::ChemicalsController

New admin API endpoints:
- GET    /admin/chemicals
- GET   /admin/chemicals/:chemical_id
- GET   /admin/chemicals/:chemical_id/map
- POST  /admin/chemicals/:chemical_id/map/:gn_id
- PATCH /admin/chemicals/:chemical_id/map/:gn_id
- PUT   /admin/chemicals/:chemical_id/map/:gn_id

Trelllo: https://trello.com/c/8WjkHQlC/868-admin-ui-to-view-edit-and-add-cas-number-mappings

Commit:
Refactor

Commit:
Refactor to support admin API:
- refactor errors in controller
- add attribute `goods_nomenclature_map` to JSON response
- update specs

Commit:
Refactor controller

alexey-klimuk commented:
@mcumcu I think you forgot to add routes

alexey-klimuk commented on app/controllers/api/admin/chemicals_controller.rb:
```diff
@@ -0,0 +1,142 @@
+          @errors << "Chemical id not found: chemical_id: #{id}"
+        end
+
+        @errors << "Chemical not found: chemical_id: #{params[:chemical_id]}" unless @chemical = Chemical.find(id: id)
```
we can use `where(id: id).take`, if record is not found we will go [there](https://github.com/bitzesty/trade-tariff-backend/blob/master/app/controllers/api/api_controller.rb#L9)

mcumcu replied:
now using `.take`

alexey-klimuk commented on app/controllers/api/admin/chemicals_controller.rb:
```diff
@@ -0,0 +1,142 @@
+      end
+      
+      def fetch_map
+        @map = ChemicalsGoodsNomenclatures.find(
```
same `where(cond).take` here

mcumcu replied:
In some cases, this method is expected to set `@map` to nil (and continue without returning a 404), so cant' use  `.take` here.

alexey-klimuk commented on app/controllers/api/admin/chemicals_controller.rb:
```diff
@@ -0,0 +1,142 @@
+        ) unless @errors.any?
+      end
+      
+      def fetch_commodity
```
we can reuse [this one](https://github.com/bitzesty/trade-tariff-backend/blob/master/app/controllers/api/admin/commodities_controller.rb#L13)

mcumcu replied:
Reusing code from `Api::Admin::CommoditiesController::find_commodity`

alexey-klimuk commented:
@mcumcu should we have endpoint to delete mapping?

alexey-klimuk commented:
@matthewford should we have ability to define new chemicals (name + cas number) manually via admin?

matthewford commented:
> @matthewford should we have ability to define new chemicals (name + cas number) manually via admin?

I think so, and delete - I do not expect it to happen often, but it might be needed once in a while, more often I think they will want to do a bulk delete/replace import.

alexey-klimuk commented:
@mcumcu @matthewford  maybe we can think about better routes/actions e.g.:

```
# to display all chemicals
GET /admin/chemicals

# to display chemical with all related goods_nomenclatures
GET /admin/chemicals/:chemical_id

# to link new goods_nomenclature with chemical
POST /admin/chemicals/:chemical_id/goods_nomenclatures/:gn_id

# to update link between goods_nomenclature and chemical
PATCH /admin/chemicals/:chemical_id/goods_nomenclatures/:gn_id
PUT /admin/chemicals/:chemical_id/goods_nomenclatures/:gn_id

# to delete link between goods_nomenclature and chemical
DELETE /admin/chemicals/:chemical_id/goods_nomenclatures/:gn_id

** gn_id - is a `goods_nomenclature_sid` OR `goods_nomenclature_item_id`
```

Commit:
Adding new routes

Commit:
Refactor
- use `.take`
- reuse code in `Api::Admin::CommoditiesController::find_commodity`

Commit:
Add `delete` endpoint

matthewford commented on app/controllers/api/admin/chemicals_controller.rb:
```diff
@@ -60,6 +60,14 @@ def update_map
+      # DELETE /admin/chemicals/:chemical_id/map/:gn_id
+      def delete_map
+        status = @map.destroy ? :ok : :not_found
+        # status = true ? :ok : :not_found
```
does this comment need to be here?

mcumcu replied:
removed the unneeded comment

matthewford reviewed (commented):
> @mcumcu @matthewford maybe we can think about better routes/actions e.g.:
> 
> ```
> # to display all chemicals
> GET /admin/chemicals
> 
> # to display chemical with all related goods_nomenclatures
> GET /admin/chemicals/:chemical_id
> 
> # to link new goods_nomenclature with chemical
> POST /admin/chemicals/:chemical_id/goods_nomenclatures/:gn_id
> 
> # to update link between goods_nomenclature and chemical
> PATCH /admin/chemicals/:chemical_id/goods_nomenclatures/:gn_id
> PUT /admin/chemicals/:chemical_id/goods_nomenclatures/:gn_id
> 
> # to delete link between goods_nomenclature and chemical
> DELETE /admin/chemicals/:chemical_id/goods_nomenclatures/:gn_id
> 
> ** gn_id - is a `goods_nomenclature_sid` OR `goods_nomenclature_item_id`
> ```

 I agree

alexey-klimuk commented on app/controllers/api/admin/commodities_controller.rb:
```diff
@@ -10,13 +10,17 @@ def show
         raise Sequel::RecordNotFound if @commodity.goods_nomenclature_item_id.in? HiddenGoodsNomenclature.codes
       end
+
+      def self.find_commodity_by_code(code)
```
@mcumcu can we do it as a model method rather than controller method?

mcumcu replied:
This class method has been moved to the Commodity model

Commit:
Add Chemical create and update routes
- add two new routes
- add specs
- add validations to Chemical and ChemicalName models
- make `find_commodity_by_code(code)` a model mothod on Commodity and respond to PR comments

mcumcu commented:
> > @mcumcu @matthewford maybe we can think about better routes/actions e.g.:
> > ... [old routes]
> 
> I agree

Trade Tariff Backend
Api::Admin::ChemicalsController

Chemical routes:
```
GET    /admin/chemicals                                                     api/admin/chemicals#index
POST   /admin/chemicals                                                     api/admin/chemicals#create
GET    /admin/chemicals/:chemical_id                                        api/admin/chemicals#show
PATCH  /admin/chemicals/:chemical_id(/:chemical_name_id/:new_chemical_name) api/admin/chemicals#update
PUT    /admin/chemicals/:chemical_id(/:chemical_name_id/:new_chemical_name) api/admin/chemicals#update
** not implemented ** DELETE /admin/chemicals/:chemical_id                  api/admin/chemicals#delete
```

Chemical <-> GN mapping routes:
```
GET    /admin/chemicals/:chemical_id/map                                    api/admin/chemicals#show_map
POST   /admin/chemicals/:chemical_id/map/:gn_sid                            api/admin/chemicals#create_map
PATCH  /admin/chemicals/:chemical_id/map/:gn_iid <- use item_id (code)      api/admin/chemicals#update_map
PUT    /admin/chemicals/:chemical_id/map/:gn_iid <- use item_id (code)      api/admin/chemicals#update_map
DELETE /admin/chemicals/:chemical_id/map/:gn_sid                            api/admin/chemicals#delete_map
```
Notes:
- `api/admin/chemicals#delete` is not (yet) implemented
- `api/admin/chemicals#update_map` - accepts the GN code (i.e., the `goods_nomenclature_item_id`) in order to support the ability to directly enter the GN code in a text input field in the Admin app. The expected parameter name is now `gn_iid`.

  The endpoint could be changed to accept the `goods_nomenclature_sid` instead, and better conform to the Rails/API routing convention of using the object's primary key as the identifier, but on the Admin UI, the user would have to:
    1) enter the `goods_nomenclature_sid` in a text field, and/or
    2) select the GN from a drop-down select field, or
    3) could also implement a JS lookup-as-you-type feature on the text field or select field

- the other endpoints where we accept a `goods_nomenclature_sid`, `#create_map` and `#delete_map`, now expect the parameter `gn_sid`. These endpoints could be changed to accept `goods_nomenclature_iid` -- but the default behaviour of the Admin app forms would have to be changed to use `goods_nomenclature_iid` as the primary key.

matthewford commented:
Is that the convention we use in other places I would just use the full param name as that is confusing otherwise

Commit:
Refactor routes

mcumcu commented:
> Is that the convention we use in other places I would just use the full param name as that is confusing otherwise

I'll use full param names

Commit:
Use model attributes for API parameter names:
- :gn_iid -> :goods_nomenclature_item_id
- :gn_sid -> :goods_nomenclature_sid

alexey-klimuk commented on app/controllers/api/admin/chemicals_controller.rb:
```diff
@@ -0,0 +1,236 @@
+
+      def fetch_map
+        unless @errors.any?
+          @map = if @commodity.nil?
```
```
@map = if @commodity.present?
  ChemicalsGoodsNomenclatures.find(
    chemical_id: @chemical.id,
    goods_nomenclature_sid: @commodity.id
  )
end
```

mcumcu replied:
Updated. Thanks for the refactoring.

alexey-klimuk commented on config/routes.rb:
```diff
@@ -36,6 +36,18 @@
+        put     '(:chemical_name_id/:new_chemical_name)', to: 'chemicals#update'
+       #delete  '',                                       to: 'chemicals#delete'
+        get     'map',                                    to: 'chemicals#show_map'
+        post    'map/:goods_nomenclature_sid',            to: 'chemicals#create_map'
```
@mcumcu why do we use both `goods_nomenclature_sid` and `goods_nomenclature_item_id ` ?

mcumcu replied:
I've refactored the routes. The new ones do use both item_id and sid. Here's why:
- `post    'map/:goods_nomenclature_item_id'` - The endpoint could be changed to accept the goods_nomenclature_sid instead, and better conform to the Rails/API routing convention of using the object's primary key as the identifier, but on the Admin UI, the user would have to:
  1.  enter the goods_nomenclature_sid in a text field, and/or
  2.  select the GN from a drop-down select field, or
  3.  could also implement a JS lookup-as-you-type feature on the text field or select field
- `patch|put   'map/:goods_nomenclature_sid(/:goods_nomenclature_item_id)'` - sid is used for the existing mapping that is to be updated, the new commodity to map to is identified by item_id, since that's what the user enters in the admin UI (see above for other options)

Commit:
Support `chemical_names`
- name* methods are needed by the Admin app
- refactor
- update specs

Commit:
Refactor based on code review

Commit:
Update Rake chemical:import task to accept an argument
- pass in a filename that will be imported by this task
-  e.g. `bundle exec rake chemical:import Annex_3.csv`
- the file to be imported must be in the `db/` directory

alexey-klimuk commented:
@matthewford should we mace cas `auditable` (same as search_references) ?

matthewford commented:
If that means they follow time travel rules that might be useful, but I'm not sure you can search for cas with a point in time atm

alexey-klimuk commented:
@matthewford  it's about keeping record changes after update

matthewford commented:
Yeah might be useful

On Tue, 24 Nov 2020 at 19:43, Alexey Klimuk <notifications@github.com>
wrote:

> @matthewford <https://github.com/matthewford> it's about keeping record
> changes after update
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitzesty/trade-tariff-backend/pull/328#issuecomment-733193498>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAABV2XTDWU3PJZ6LOX3R3SRQEGTANCNFSM4TJTYZHA>
> .
>
-- 
Sent from my iPhone

-- 
Bit Zesty Ltd, a company incorporated in England & Wales with registered 
number 06883289. Our registered address is 94 White Lion Street, London, N1 
9PF.

alexey-klimuk commented on app/controllers/api/admin/chemicals_controller.rb:
```diff
@@ -0,0 +1,238 @@
+
+      # GET    /admin/chemicals
+      def index
+        respond_with Chemical.order_by(:id).all
```
@mcumcu could you please add pagination?

mcumcu replied:
Pagination added to `/admin/chemicals`
commit: 4967b6eb

alexey-klimuk commented on app/controllers/api/admin/chemicals_controller.rb:
```diff
@@ -0,0 +1,238 @@
+            @chemical.update(cas: params[:cas])
+          end
+
+          if params[:chemical_name_id].present?
```
@mcumcu no need to check if param is present, it's already done in L48

mcumcu replied:
L48 checks if either param is present. Here we need to check specifically for `params[: chemical_name_id]`, and update if it exists

alexey-klimuk commented on app/controllers/api/admin/chemicals_controller.rb:
```diff
@@ -0,0 +1,238 @@
+        end
+
+        Sequel::Model.db.transaction do
+          if params[:cas].present?
```
@mcumcu no need to check if param is present, it's already done in L48

mcumcu replied:
L48 checks if either param is present. Here we need to check specifically for `params[:cas]`, and update if it exists

alexey-klimuk commented on app/controllers/api/admin/chemicals_controller.rb:
```diff
@@ -0,0 +1,238 @@
+module Api
```
@mcumcu this controller looks a bit overloaded with multiple responsibilities. I think we can move some validations on model or db level.

Commit:
Remove comment

Commit:
Paginate `/admin/chemicals`

Commit:
Add pagination to `/chemicals`

Commit:
Make Chemical auditable
- add auditable plugin to Chemical model
- add spec

mcumcu commented:
> Yeah might be useful
>> On Tue, 24 Nov 2020 at 19:43, Alexey Klimuk ***@***.***> wrote: @matthewford <https://github.com/matthewford> it's about keeping record changes after update — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#328 (comment)](https://github.com/bitzesty/trade-tariff-backend/pull/328#issuecomment-733193498)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAABV2XTDWU3PJZ6LOX3R3SRQEGTANCNFSM4TJTYZHA> .

Added auditable plugin to Chemical model

Commit:
Refactor ChemicalsController
- move code from the controller into several models

Commit:
Change HTTP response code to 200 OK for chemical patch/put operations on admin API
- was 202 Accepted, now 200 OK

matthewford commented:
Engine will merge
